time-it-builder/src/lib.rs

```diff
@@ -37,37 +77,123 @@ impl TimeItBuilder {
-        let mut config = TIMER_CONFIG.write();
+        let mut config = TIMEIT_CONFIG.write();
         config.action = action;
         self
     }
 
     pub fn default_tag(&self, tag: impl ToString) -> &Self {
-        let mut config = TIMER_CONFIG.write();
+        let mut config = TIMEIT_CONFIG.write();
         config.default_tag = tag.to_string();
         self
     }
+
+    pub fn mode(&self, mode: TimeItMode) -> &Self {
+        let mut config = TIMEIT_CONFIG.write();
+        config.mode = mode;
+        self
+    }
 }
 
 pub fn get_config() -> &'static TimeItConfig {
-    LAZY_TIMER_CONFIG.get_or_create(|| TIMER_CONFIG.read().clone())
+    LAZY_TIMEIT_CONFIG.get_or_create(|| {
+        let mut config = TIMEIT_CONFIG.write();
+        let mut lazy_config = std::mem::replace(&mut *config, TimeItConfig::default());
+        let default_tag = lazy_config.default_tag.clone();
+        let shutdown = config.shutdown.clone();
+        match lazy_config.mode {
+            TimeItMode::Async => {
+                let (sender, receiver) = bounded::<TimeItEvent>(100);
+                let handle = thread::spawn(move || {
+                    while !(shutdown.load(Ordering::Relaxed) && receiver.is_empty()) {
+                        match receiver.recv() {
+                            Ok(event) => {
+                                (lazy_config.action)(
+                                    event.tag.as_ref().unwrap_or(&default_tag),
+                                    event.duration,
+                                );
+                            }
+                            Err(_) => {
+                                break;
+                            }
+                        }
+                    }
+                });
+                lazy_config.sender = Some(sender);
+                // Keep the config in the original struct because we want to get mutable reference to it later for cleaning up
+                config.receiver_handle = Some(handle);
+                lazy_config
+            }
+            TimeItMode::Sync => lazy_config,
+        }
+    })
+}
+
+pub fn run(tag: Option<&str>, duration: Duration) {
+    let config = get_config();
+    match config.mode {
+        TimeItMode::Async => {
+            emit(tag, duration);
+        }
+        TimeItMode::Sync => {
+            action(tag, duration);
+        }
+    }
+}
+
+pub fn wait_for_finish() {
+    let mut config = TIMEIT_CONFIG.write();
+    config.shutdown.store(true, Ordering::Relaxed);
+    config.wait_for_finish()
 }
 
 pub fn action(tag: Option<&str>, duration: Duration) {
     let config = get_config();
     (config.action)(tag.unwrap_or(&config.default_tag), duration);
 }
 
+pub fn emit(tag: Option<&str>, duration: Duration) {
+    let config = get_config();
+    let sender = config.sender.as_ref().unwrap();
+    sender
+        .send(TimeItEvent {
+            tag: tag.map(|s| s.to_string()),
+            duration,
+        })
+        .unwrap();
+}
+
```

Comment:
since run is now the function called from the macro, we should make both of these private. Don't want a user to be able to call emit themselves when they are in sync mode since this will panic in that case

Reply:
yes ok